flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -194,23 +193,28 @@ public boolean processInput() throws Exception {
 
 		int readingInputIndex = selectNextReadingInputIndex();
 		if (readingInputIndex == -1) {
-			return false;
+			return InputStatus.NOTHING_AVAILABLE;
 		}
 		lastReadInputIndex = readingInputIndex;
 
 		InputStatus status;
 		if (readingInputIndex == 0) {
 			status = input1.emitNext(output1);
+			firstInputStatus = status;
 		} else {
 			status = input2.emitNext(output2);
+			secondInputStatus = status;
 		}
 		checkFinished(status, lastReadInputIndex);
 
 		if (status == InputStatus.NOTHING_AVAILABLE) {
 			inputSelectionHandler.setUnavailableInput(readingInputIndex);
 		}
 
-		return status == InputStatus.MORE_AVAILABLE;
+		if (status == InputStatus.END_OF_INPUT && secondInputStatus != InputStatus.END_OF_INPUT) {
```

Comment:
I think this is not quite correct. What if `status == NOTHING_AVAILABLE` and for example `secondInputStatus == MORE_AVAILABLE`? Result should be `MORE_AVAILABLE` as long as input selection is correct. 

Similar, what about `status == END_OF_INPUT` and `secondInputStatus == MORE_AVAILABLE`? Currently that would return `NOTHING_AVAILABLE`?

Am I missing something, or is this wrong? If this is a bug, please also add a test coverage for those (and potentially other corner cases).

Reply:
I think we keep the same behavior as before. If `status == NOTHING_AVAILABLE` and `secondInputStatus == MORE_AVAILABLE`, it would return `status` and in `StreamTask#performDefaultAction` it would call `inputProcessor.isAvailable().thenRun()`. 

Before refactoring, if `status == NOTHING_AVAILABLE` which corresponds to the case of returning null `recordOrMark`, then in `StreamTask#performDefaultAction` it would still call `inputProcessor.isAvailable().thenRun()`. 

Also in previous behavior only two inputs are both finished, then we can return `status==END_OF_INPUT`. If only current `status== END_OF_INPUT` then the returned `recordOrMark` should be null, so in StreamTask#performDefaultAction` it would also enter the path of `inputProcessor.isAvailable().thenRun()`. In order to keep this same behavior we need to return `NOTHING_AVAILABLE` here to enter the same path in `StreamTask`.

So this part was my main concern before, whether it is worth doing this refactor here. Because we add this condition `if (status == InputStatus.END_OF_INPUT && secondInputStatus != InputStatus.END_OF_INPUT)` to always check in every call for keeping the previous behavior.

Reply:
Hmmm, maybe we should clean this up?

I think previously it was not that bad, as it was more of an optimisation, to avoid checking `isAvailable()` when hot looping (touching `isAvailable()` means touching `volatile` field inside `CompletableFuture`). 

Now that we want to actually express `END_OF_INPUT` as well, we might need to be more precise here.

Do you think it would be difficult to fix this logic here?

Reply:
Let me think how to make this logic more clearly now.